src/main/java/hudson/remoting/Engine.java

```diff
@@ -801,38 +807,115 @@ public void closeRead() throws IOException {
                 if (noReconnect) {
                     return;
                 }
-                firstAttempt = Instant.now();
                 events.onDisconnect();
-                while (true) {
-                    // TODO refactor various sleep statements into a common method
-                    if (Util.shouldBailOut(firstAttempt, noReconnectAfter)) {
-                        events.status("Bailing out after " + DurationFormatter.format(noReconnectAfter));
-                        return;
-                    }
-                    TimeUnit.SECONDS.sleep(10);
-                    // Unlike JnlpAgentEndpointResolver, we do not use $jenkins/tcpSlaveAgentListener/, as that will be
-                    // a 404 if the TCP port is disabled.
-                    URL ping = new URL(hudsonUrl, "login");
-                    try {
-                        HttpURLConnection conn = (HttpURLConnection) ping.openConnection();
-                        int status = conn.getResponseCode();
-                        conn.disconnect();
-                        if (status == 200) {
-                            break;
-                        } else {
-                            events.status(ping + " is not ready: " + status);
-                        }
-                    } catch (IOException x) {
-                        events.status(ping + " is not ready", x);
-                    }
-                }
                 reconnect();
             }
         } catch (Exception e) {
             events.error(e);
         }
     }
 
+    /**
+     * Evaluates a condition with exponential backoff until it succeeds or the timeout is reached.
+     * @param condition the condition to attempt to succeed with exponential backoff
+     * @return true if the condition succeeded, false if the condition failed and the timeout was reached
+     * @throws InterruptedException if the thread was interrupted while waiting.
+     */
+    private boolean succeedsWithRetries(SupplierThrowingException<Boolean> condition) throws InterruptedException {
+        var exponentialRetry = new ExponentialRetry(noReconnectAfter);
+        while (exponentialRetry != null) {
+            try {
+                if (condition.get()) {
+                    return true;
+                }
+            } catch (Exception x) {
+                events.status("Failed to connect: " + x.getMessage());
```

Comment:
We are losing the stack trace here; is that potentially important for diagnosis?

Reply:
In my experience these stacktraces are noisy. I could eventually log them with standard JUL logger (FINE) in parallel, but I don't really see the point of keeping them in status.